File: android/src/com/boliao/eod/WeatherRepo.java
package com.boliao.eod;

//import androidx.annotation.NonNull;
//import androidx.lifecycle.MutableLiveData;
import android.app.Application;
import android.support.annotation.NonNull;
import android.arch.lifecycle.MutableLiveData;
import android.util.Log;

import com.android.volley.Cache;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.DiskBasedCache;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.awt.font.TextAttribute;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

/**
 * Where most of the hard work gets done... at the lower levels...
 */
public class WeatherRepo {
    private static final String TAG = "WeatherRepo";

    // Mocking var
    private static int count = 0;

    // interval between fetching data
    private static final int FETCH_INTERVAL_MILLIS = 2000;

    // threading
    private Runnable weatherRunner;

    // weather live data (writable)
    @NonNull
    private MutableLiveData<String> weatherData = new MutableLiveData<>();
    @NonNull
    MutableLiveData<String> getWeatherData() {
        return weatherData;
    }

    // singleton pattern boilerplate
    private static WeatherRepo instance;
    static WeatherRepo getInstance() {
        if (instance == null) {
            synchronized (WeatherRepo.class) {
                if(instance == null)
                    instance = new WeatherRepo();
            }
        }
        return instance;
    }

    /**
     * Mock live data.
     */
    void mockOnlineWeatherData() {
        weatherData.postValue("Mock Weather Data");
    }

    /**
     * TODO THREADING 2: override method to fetch mock timed weather data
     * - background continuous task to fetch mock weather data
     * - always updating regularly (confirm < 15min) from online API
     * - not expecting to pause it at any point
     * - ideally want updates even if navigate away
     * - Q: what primitive should we use?
     * - Recurring WorkManager?
     * - IntentService?
     * - ThreadPoolExecutor?
     * - AsyncTask?
     * - A: Spawn a HandlerThread
     * - Q: Should you use a service to wrap the thread?
     * - A: Depends on whether you want it running beyond visible lifecycle
     */
    void fetchMockOnlineWeatherData() {
        final WeatherWorkerThread weatherWorkerThread = new WeatherWorkerThread();
        weatherWorkerThread.start();
        weatherWorkerThread.prepareHandler();

        weatherRunner = new Runnable() {
            @Override
            public void run() {
                weatherData.postValue("Weather now is: " + count++);
                weatherWorkerThread.postTaskDelayed(weatherRunner, FETCH_INTERVAL_MILLIS);
            }
        };
        weatherWorkerThread.postTask(weatherRunner);
    }

    /**
     * TODO NETWORKING 1: fetch real online weather data from RESTful API
     * - volley is fast and clean
     * - many APIs were created pre-MVVM, so needed context to init, in this case the request queue
     * - note how I've used the semi singleton to resolve the context issue
     * - others have used empty Application classes
     */
    void fetchOnlineWeatherData() {
        final String urlStr = "https://api.data.gov.sg/v1/environment/2-hour-weather-forecast?date=" +
                getToday();
        Log.i(TAG, "Fetching online weather data: url=" + urlStr);

        // form the network request complete with response listener
        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, urlStr, null, new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {
                Log.i(TAG, "volley fetched \n" + response.toString());
                try {
                    // parse the returned json
                    String forecastStr = response.getJSONArray("items")
                            .getJSONObject(0)
                            .getJSONArray("forecasts")
                            .getJSONObject(0)
                            .getString("forecast");

                    // post to live data here
                    weatherData.postValue("Weather now is " + forecastStr);
                } catch (JSONException e) {
                    Log.e(TAG, "json exception: " + e.getLocalizedMessage());
                }
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                Log.e(TAG, "Volley error while fetching :" + error.getLocalizedMessage());
            }
        });

        NetworkRequestQueue.getInstance().add(request);
    }

    /**
     * Helper to get today's date in API format for network request.
     * @return
     */
    String getToday() {
        Date time = Calendar.getInstance().getTime();
        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH);
        return dateFormat.format(time);
    }
}

File: android/src/com/boliao/eod/WeatherRepo.kt
package com.boliao.eod

import android.util.Log
import androidx.lifecycle.MutableLiveData
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import kotlinx.coroutines.*
import org.json.JSONException
import java.lang.Runnable
import java.text.SimpleDateFormat
import java.util.*

/**
 * The Weather Repository that manages weather data.
 * @param networkRequestQueue is the Volley component that manages network requests.
 */
class WeatherRepo (private val networkRequestQueue: NetworkRequestQueue) {

    companion object {
        private val TAG = WeatherRepo::class.simpleName

        // interval between fetching data
        private const val FETCH_INTERVAL_MILLIS: Long = 1000
    }

    // mocking var
    private var count = 0

    // threading
    private lateinit var weatherRunner: Runnable

    // weather live data (writable)
    val weatherData = MutableLiveData<String>()

    // TODO THREADING: observe the setup of some basic networking arch
    // For the lib, we'll use volley as example here (key advantages: fast and clean)
    // 1. define some permissions in the manifest
    // 2. create NetWorkRequestQueue singleton
    // 3. in EODApp, init and link up the NetWorkRequestQueue with the App context
    // 4. in EODApp, init a weatherRepo with the NetworkRequestQueue
    // 5. form a Volley request with the URL
    //    - define a Response.Listener to handle the Response
    //    - define a handler to handle errors

    private val urlStr = "https://api.data.gov.sg/v1/environment/2-hour-weather-forecast?date_time=$today"
    private val request = JsonObjectRequest(
            Request.Method.GET,
            urlStr,
            null,
            { response ->
                // Log.i(TAG, "volley fetched \n$response")
                try { // parse the returned json
                    val areaStr = response.getJSONArray("items")
                            .getJSONObject(0)
                            .getJSONArray("forecasts")
                            .getJSONObject(0)
                            .getString("area")
                    val forecastStr = response.getJSONArray("items")
                            .getJSONObject(0)
                            .getJSONArray("forecasts")
                            .getJSONObject(0)
                            .getString("forecast")

                    // post to live data here
                    weatherData.postValue("Weather in $areaStr \nat\n${today.substring(11, 19)} \nis \n$forecastStr  ")

                } catch (e: JSONException) {
                    Log.e(TAG, "json exception: " + e.localizedMessage)
                }
            },
            { error ->
                Log.e(TAG, "Volley error while fetching :" + error.localizedMessage)
            }
    )

    /**
     * OPTIONAL TODO THREADING 3: override method to fetch timed weather data
     * - background continuous task to fetch weather data
     * - always updating regularly (confirm < 15min) from online API
     * - not expecting to pause it at any point
     * - ideally continue to updates as much as possible even if navigate away
     *
     * Q: what primitive should we use?
     * - Recurring WorkManager?
     * - IntentService?
     * - ThreadPoolExecutor?
     * - AsyncTask?
     * A: Spawn a HandlerThread
     * Q: Should you use a service to contain the thread?
     * A: Depends on whether you want it running beyond visible lifecycle
     *
     * - create a WeatherWorkerThread from HandlerThread class
     * - start the thread and prepare the handler (looper only available after init HandlerThread)
     * - create a Runnable to postValue to the weatherData (simply post a counter value)
     * - post the Runnable as a delayed task in the thread to time updates
     * - goto SplashViewModel for THREADING 4
     *
     *
     * - use the HandlerThread primitive to make timed requests to the web API
     * - goto SplashViewModel for NETWORKING 3
     * NOTE that posting to the Livedata is done in the volley callback
     */
    fun fetchOnlineWeatherData() {
        val weatherWorkerThread = WeatherWorkerThread()
        weatherWorkerThread.start()
        weatherWorkerThread.prepareHandler()
        weatherRunner = Runnable {
            // DEBUG mocking weather with count
            // weatherData.postValue("Weather now is: ${count++}")

            // add request (with it's async response) to the volley queue
            networkRequestQueue.add(request)
            weatherWorkerThread.postTaskDelayed(weatherRunner, FETCH_INTERVAL_MILLIS.toLong())
        }
        weatherWorkerThread.postTask(weatherRunner)
    }

    /**
     * TODO THREADING 5: use a coroutine to fetch the weather
     * 1. create a suspend fun that dispatches to the IO threadpool
     * 2. make an infinite loop adds a volley request to the queue regularly
     *
     * NOTE that Repo can't start any coroutines since it does not have a LifeCycle to manage it
     *      - but it can certainly provide the suspend function (which we are not using here)
     *      - if threading needs to be handled here, it is a perfect use case for low level handlerthreads
     */

    /**
     * Helper to get today's date in API format for network request.
     * @return String in the format that the data.gov.sg API likes
     */
    val today: String
        get() {
            val time = Calendar.getInstance().time
            val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.ENGLISH)
            return dateFormat.format(time)
        }
}
